gossipdns/nameserver.go

```diff
@@ -0,0 +1,234 @@
+	n.RLock()
+	defer n.RUnlock()
+
+	entries := n.entries.lookup(hostname)
+	result := []address.Address{}
+	for _, e := range entries {
+		if e.Tombstone > 0 {
+			continue
+		}
+		result = append(result, e.Addr)
+	}
+	return result
+}
+
+func (n *Nameserver) ReverseLookup(ip address.Address) (string, error) {
+	n.RLock()
+	defer n.RUnlock()
+
+	match, err := n.entries.first(func(e *Entry) bool {
+		return e.Addr == ip
+	})
+	if err != nil {
+		return "", err
+	}
+	return match.Hostname, nil
+}
+
+func (n *Nameserver) ContainerDied(ident string) error {
+	n.Lock()
+	defer n.Unlock()
+	n.entries.tombstone(n.ourName, func(e *Entry) bool {
+		if e.ContainerID == ident {
+			Log.Infof("[gossipdns] container %s died, tombstoning entry %s", ident, e.String())
+			return true
+		}
+		return false
+	})
+	return nil
+}
+
+func (n *Nameserver) PeerGone(peer *router.Peer) {
+	n.Lock()
+	defer n.Unlock()
+	n.entries.delete(func(e *Entry) bool {
+		return e.Origin == peer.Name
+	})
+}
+
+func (n *Nameserver) Delete(hostname, containerid, ipStr string, ip address.Address) {
+	n.Lock()
+	defer n.Unlock()
+	Log.Infof("[gossipdns] tombstoning hostname=%s, containerid=%s, ip=%s", hostname, containerid, ipStr)
+	n.entries.tombstone(n.ourName, func(e *Entry) bool {
+		if hostname != "*" && e.Hostname != hostname {
+			return false
+		}
+
+		if containerid != "*" && e.ContainerID != containerid {
+			return false
+		}
+
+		if ipStr != "*" && e.Addr != ip {
+			return false
+		}
+		return true
+	})
+}
+
+func (n *Nameserver) deleteTombstones() {
+	n.Lock()
+	defer n.Unlock()
+	now := time.Now().Unix()
+	n.entries.delete(func(e *Entry) bool {
+		return now-e.Tombstone > int64(tombstoneTimeout/time.Second)
+	})
+}
+
+func (n *Nameserver) String() string {
+	n.RLock()
+	defer n.RUnlock()
+	var buf bytes.Buffer
+	for _, entry := range n.entries {
+		if entry.Tombstone > 0 {
+			continue
+		}
+		containerid := entry.ContainerID
+		if len(containerid) > 12 {
+			containerid = containerid[:12]
+		}
+		fmt.Fprintf(&buf, "%s: %s [%s]\n", containerid, entry.Hostname, entry.Addr.String())
+	}
+	return buf.String()
+}
+
+func (n *Nameserver) Gossip() router.GossipData {
+	n.RLock()
+	defer n.RUnlock()
+	result := make(Entries, len(n.entries))
+	copy(result, n.entries)
+	return &result
+}
+
+func (n *Nameserver) OnGossipUnicast(sender router.PeerName, msg []byte) error {
+	return nil
+}
+
+func (n *Nameserver) receiveGossip(msg []byte) (router.GossipData, router.GossipData, error) {
+	var entries Entries
+	if err := gob.NewDecoder(bytes.NewReader(msg)).Decode(&entries); err != nil {
+		return nil, nil, err
+	}
+
+	if err := entries.check(); err != nil {
+		return nil, nil, err
+	}
+
+	// Filter out entries which we think have died.  Have to be
+	// careful to avoid a deadlock - other threads can take the
+	// peers write lock, then call out OnGC callback, taking our
+	// write lock.
+	if n.peers != nil {
+		n.peers.RLock()
+		defer n.peers.RUnlock()
```

Comment:
`FetchUnlocked` - gross. Could the `OnGC` callback take place outside the lock? If it can then it definitely should; invoking callbacks while holding locks is bad.